sdks/python/apache_beam/io/vcfio.py

```diff
@@ -0,0 +1,428 @@
+
+  def __init__(self,
+               file_pattern,
+               compression_type=CompressionTypes.AUTO,
+               buffer_size=DEFAULT_VCF_READ_BUFFER_SIZE,
+               validate=True):
+    super(_VcfSource, self).__init__(file_pattern,
+                                     compression_type=compression_type,
+                                     validate=validate)
+
+    self._header_lines_per_file = {}
+    self._file_pattern = file_pattern
+    self._compression_type = compression_type
+    self._buffer_size = buffer_size
+    self._validate = validate
+
+  def read_records(self, file_name, range_tracker):
+    record_iterator = _VcfSource._VcfRecordIterator(
+        file_name,
+        range_tracker,
+        self._file_pattern,
+        0,  # min_bundle_size
+        self._compression_type,
+        True,  # strip_trailing_newlines
+        coders.StrUtf8Coder(),  # coder
+        buffer_size=self._buffer_size,
+        validate=self._validate,
+        skip_header_lines=0)
+
+    # Convert iterator to generator to abstract behavior
+    for line in record_iterator:
+      yield line
+
+  class _VcfRecordIterator(object):
+    """An Iterator for processing a single VCF file."""
+
+    def __init__(self,
+                 file_name,
+                 range_tracker,
+                 file_pattern,
+                 min_bundle_size,
+                 compression_type,
+                 strip_trailing_newlines,
+                 coder,
+                 **kwargs):
+      self._header_lines = []
+      self._last_record = None
+      self._file_name = file_name
+
+      text_source = TextSource(
+          file_pattern,
+          min_bundle_size,
+          compression_type,
+          strip_trailing_newlines,
+          coder,
+          header_processor_fns=(lambda x: x.startswith('#'),
+                                self._store_header_lines),
+          **kwargs)
+
+      self._text_lines = text_source.read_records(self._file_name,
+                                                  range_tracker)
+      try:
+        self._vcf_reader = vcf.Reader(fsock=self._create_generator())
+      except SyntaxError as e:
+        raise ValueError('Invalid VCF header %s' % str(e))
+
+    def _store_header_lines(self, header_lines):
+      self._header_lines = header_lines
+
+    def _create_generator(self):
+      header_processed = False
+      for record in self._text_lines:
+        if not header_processed and self._header_lines:
+          for header in self._header_lines:
+            self._last_record = header
+            yield self._last_record
+          header_processed = True
+        # PyVCF has explicit str() calls when parsing INFO fields, which fails
+        # with UTF-8 decoded strings. Encode the line back to UTF-8.
+        self._last_record = record.encode('utf-8')
+        yield self._last_record
+
+    def __iter__(self):
+      return self
+
+    def next(self):
+      try:
+        record = next(self._vcf_reader)
```

Comment:
<!--new_thread; commit:63511ddb8caf9e5c12888ef8a4d1071e0ea6c168; resolved:0-->
Does vcf.Reader raise StopIteration at the end ?

Reply:
Yes